sns/src/main/scala/akka/stream/alpakka/sns/SnsPublishSinkStage.scala

```diff
@@ -0,0 +1,77 @@
+package akka.stream.alpakka.sns
+
+import akka.Done
+import akka.stream.stage._
+import akka.stream.{Attributes, Inlet, SinkShape}
+import com.amazonaws.handlers.AsyncHandler
+import com.amazonaws.services.sns.AmazonSNSAsync
+import com.amazonaws.services.sns.model.{PublishRequest, PublishResult}
+
+import scala.concurrent.{Future, Promise}
+
+final class SnsPublishSinkStage(topicArn: String, snsClient: AmazonSNSAsync)
+    extends GraphStageWithMaterializedValue[SinkShape[String], Future[Done]] {
+
+  private val in = Inlet[String]("SnsPublishSink.in")
+
+  override def shape: SinkShape[String] = SinkShape.of(in)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[Done]) = {
+    val completed = Promise[Done]()
+
+    val logic = new GraphStageLogic(shape) with StageLogging {
+
+      private val failureCallback = getAsyncCallback[Throwable](handleFailure)
+      private val successCallback = getAsyncCallback[PublishResult](handleSuccess)
+
+      setHandler(in,
+        new InHandler {
+        override def onPush(): Unit = {
+          val request = new PublishRequest().withTopicArn(topicArn).withMessage(grab(in))
+
+          snsClient.publishAsync(request,
+            new AsyncHandler[PublishRequest, PublishResult] {
+            override def onError(exception: Exception): Unit =
+              failureCallback.invoke(exception)
+
+            override def onSuccess(request: PublishRequest, result: PublishResult): Unit =
+              successCallback.invoke(result)
+          })
+        }
+
+        override def onUpstreamFinish(): Unit = {
+          completeStage()
+          completed.trySuccess(Done)
+        }
+
+        override def onUpstreamFailure(ex: Throwable): Unit = {
+          log.error(ex, "Upstream failure: {}", ex.getMessage)
+          failStage(ex)
+          completed.tryFailure(ex)
+        }
+      })
+
+      override def preStart(): Unit = {
+        setKeepGoing(true)
+        pull(in)
+      }
+
+      def handleFailure(ex: Throwable): Unit = {
+        log.error(ex, "AmazonSNSAsync failure: {}", ex.getMessage)
```

Comment:
No need to log the failure since we fail the stage and the materialized value with it. You could wrap it with a en exception to decorate with extra information from the stage though.